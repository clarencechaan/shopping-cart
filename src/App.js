import { BrowserRouter, Routes, Route } from "react-router-dom";
import NavBar from "./components/NavBar";
import Home from "./components/Home";
import MusicShop from "./components/MusicShop";
import MerchShop from "./components/MerchShop";
import SaleShop from "./components/SaleShop";
import Cart from "./components/Cart";
import Details from "./components/Details";
import items from "./items/items.js";
import "./styles/App.css";
import { useState, useEffect } from "react";

const App = () => {
  const [shopItems, setShopItems] = useState([]);
  const [cart, setCart] = useState([]);

  useEffect(() => setShopItems(items), []);

  function handleAddToCart(item, qty) {
    if (!qty) return;
    if (!isItemInCart(item, cart)) {
      setCart((prevCart) => {
        const newCartItem = { ...item, quantity: qty };
        return [...prevCart, newCartItem];
      });
    } else {
      setCart((prevCart) => {
        const prevCartItem = prevCart.find((c) => c.id === item.id);
        const newCartItem = {
          ...prevCartItem,
          quantity: prevCartItem.quantity + qty,
        };
        const index = prevCart.findIndex((c) => c.id === item.id);
        return [
          ...prevCart.slice(0, index),
          newCartItem,
          ...prevCart.slice(index + 1),
        ];
      });
    }
    removeZeroQuantity();
  }

  function removeZeroQuantity() {
    setCart((prevCart) => prevCart.filter((item) => item.quantity > 0));
  }

  function isItemInCart(item) {
    for (const c of cart) {
      if (c.id === item.id) return true;
    }
    return false;
  }

  function getQtyInCart(id) {
    if (isItemInCart({ id: id })) {
      return cart.find((c) => c.id === id).quantity;
    } else {
      return 0;
    }
  }

  function getTotalQtyInCart() {
    let total = 0;
    for (const c of cart) {
      total += c.quantity;
    }
    return total;
  }

  // function addQty(id, qty) {
  //   setCart((prevCart) => {
  //     const prevCartItem = prevCart.find((c) => c.id === id);
  //     const newCartItem = {
  //       ...prevCartItem,
  //       quantity: prevCartItem.quantity + 1,
  //     };
  //     const index = prevCart.findIndex((c) => c.id === id);
  //     return [
  //       ...prevCart.slice(0, index),
  //       newCartItem,
  //       ...prevCart.slice(index + 1),
  //     ];
  //   });
  // }

  console.log(cart);

  return (
    <BrowserRouter>
      <NavBar totalQtyInCart={getTotalQtyInCart()} />
      <Routes>
        <Route path="/" element={<Home />} />
        <Route
          path="/music-shop"
          element={<MusicShop shopItems={shopItems} />}
        />
        <Route
          path="/merch-shop"
          element={<MerchShop shopItems={shopItems} />}
        />
        <Route path="/sale-shop" element={<SaleShop shopItems={shopItems} />} />
        <Route
          path="/cart"
          element={
            <Cart
              cartItems={cart}
              getTotalQtyInCart={getTotalQtyInCart}
              handleAddToCart={handleAddToCart}
            />
          }
        />
        <Route
          path="/details"
          element={
            <Details
              handleAddToCart={handleAddToCart}
              getQtyInCart={getQtyInCart}
            />
          }
        />
      </Routes>
    </BrowserRouter>
  );
};

export default App;
