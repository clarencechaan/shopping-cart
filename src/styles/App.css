body {
  margin: 0;
}

#root {
  width: 100vw;
  height: 100vh;
  max-width: 100%;
  font-family: "Pavanam", sans-serif;
  display: flex;
  flex-direction: column;
  align-items: center;
}

#root > div:last-child {
  flex: 1;
}

.home {
  width: 100%;
  height: 100%;
}

.music-shop,
.merch-shop,
.sale-shop,
.cart {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  width: 1192px;
  padding: 0 24px 24px 24px;
}

.cart {
  width: 1096px;
}
