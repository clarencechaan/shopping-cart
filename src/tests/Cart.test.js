import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "../App";

describe("Cart component", () => {
  it("renders Cart", () => {
    const { container } = render(<App />);
    const cart = screen.getByAltText("cart");
    userEvent.click(cart);

    expect(container).toMatchSnapshot();
  });

  it("renders Cart with item", () => {
    const { container } = render(<App />);
    const music = screen.getByText(/^MUSIC$/);
    userEvent.click(music);

    const img = screen.getByAltText(
      "My Beautiful Dark Twisted Fantasy by Kanye West"
    );
    userEvent.click(img);

    const button = screen.getByRole("button", {
      name: /^ADD TO CART$/,
    });
    userEvent.click(button);

    const cart = screen.getByAltText("cart");
    userEvent.click(cart);

    const mbdtf = screen.getByText(
      "My Beautiful Dark Twisted Fantasy by Kanye West"
    );
    expect(mbdtf).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it("increment and decrement quantity", () => {
    const { container } = render(<App />);

    const music = screen.getByText(/^MUSIC$/);
    userEvent.click(music);

    const img = screen.getByAltText(
      "My Beautiful Dark Twisted Fantasy by Kanye West"
    );
    userEvent.click(img);

    const button = screen.getByRole("button", {
      name: /^ADD TO CART$/,
    });
    userEvent.click(button);

    const cart = screen.getByAltText("cart");
    userEvent.click(cart);

    const incrementor = screen.getByRole("button", {
      name: "+",
    });
    userEvent.click(incrementor);

    const qtyLabel = screen.getByTestId("quantity-label");
    expect(qtyLabel.textContent).toBe("2");

    const decrementor = screen.getByRole("button", {
      name: "-",
    });

    userEvent.click(decrementor);
    expect(qtyLabel.textContent).toBe("1");

    expect(container).toMatchSnapshot();
  });

  it("remove cart item", () => {
    const { container } = render(<App />);

    const music = screen.getByText(/^MUSIC$/);
    userEvent.click(music);

    const img = screen.getByAltText(
      "My Beautiful Dark Twisted Fantasy by Kanye West"
    );
    userEvent.click(img);

    const button = screen.getByRole("button", {
      name: /^ADD TO CART$/,
    });
    userEvent.click(button);

    const cart = screen.getByAltText("cart");
    userEvent.click(cart);

    const remove = screen.getByAltText("remove");
    userEvent.click(remove);

    const linkElement = screen.queryByText(
      "My Beautiful Dark Twisted Fantasy by Kanye West"
    );
    expect(linkElement).toBeNull();

    expect(container).toMatchSnapshot();
  });
});
