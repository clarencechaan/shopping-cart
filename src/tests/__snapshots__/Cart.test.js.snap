// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart component increment and decrement quantity 1`] = `
<div>
  <nav
    class="nav-bar"
  >
    <a
      class="logo"
      href="/shopping-cart"
    >
      auditty
    </a>
    <div>
      <a
        class="nav-link"
        href="/shopping-cart"
      >
        HOME
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/music-shop"
      >
        MUSIC
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/merch-shop"
      >
        MERCH
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/sale-shop"
        style="color: red;"
      >
        SALE
      </a>
    </div>
    <a
      class="cart-logo"
      href="/shopping-cart/cart"
    >
      <img
        alt="cart"
        src="shopping-cart.png"
      />
      <div
        class="cart-badge visible"
        data-testid="cart-badge"
      >
        1
      </div>
    </a>
  </nav>
  <div
    class="cart"
  >
    <h2>
      Your Cart (
      1
      )
    </h2>
    <div
      class="cart-items-container"
    >
      <div
        class="cart-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt=""
            class="cart-item-img"
            src="mbdtf.jpg"
          />
        </a>
        <div
          class="quantity-control"
        >
          <button
            type="button"
          >
            -
          </button>
          <div
            class="quantity-label"
            data-testid="quantity-label"
          >
            1
          </div>
          <button
            type="button"
          >
            +
          </button>
          <img
            alt="remove"
            class="remove"
            src="trash.png"
          />
        </div>
        <div
          class="cart-title"
        >
          My Beautiful Dark Twisted Fantasy by Kanye West
        </div>
        <div
          class="cart-price"
        >
          $
          11.99
        </div>
      </div>
    </div>
    <div
      class="total"
    >
      <img
        alt=""
        class="total-cart-logo"
        src="shopping-cart-white.png"
      />
      $
      11.99
    </div>
  </div>
</div>
`;

exports[`Cart component remove cart item 1`] = `
<div>
  <nav
    class="nav-bar"
  >
    <a
      class="logo"
      href="/shopping-cart"
    >
      auditty
    </a>
    <div>
      <a
        class="nav-link"
        href="/shopping-cart"
      >
        HOME
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/music-shop"
      >
        MUSIC
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/merch-shop"
      >
        MERCH
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/sale-shop"
        style="color: red;"
      >
        SALE
      </a>
    </div>
    <a
      class="cart-logo"
      href="/shopping-cart/cart"
    >
      <img
        alt="cart"
        src="shopping-cart.png"
      />
      <div
        class="cart-badge hidden"
      />
    </a>
  </nav>
  <div
    class="cart"
  >
    <h2>
      Your Cart (
      0
      )
    </h2>
    <div
      class="cart-items-container"
    />
    <div
      class="total"
    >
      <img
        alt=""
        class="total-cart-logo"
        src="shopping-cart-white.png"
      />
      $
      0.00
    </div>
  </div>
</div>
`;

exports[`Cart component renders Cart 1`] = `
<div>
  <nav
    class="nav-bar"
  >
    <a
      class="logo"
      href="/shopping-cart"
    >
      auditty
    </a>
    <div>
      <a
        class="nav-link"
        href="/shopping-cart"
      >
        HOME
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/music-shop"
      >
        MUSIC
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/merch-shop"
      >
        MERCH
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/sale-shop"
        style="color: red;"
      >
        SALE
      </a>
    </div>
    <a
      class="cart-logo"
      href="/shopping-cart/cart"
    >
      <img
        alt="cart"
        src="shopping-cart.png"
      />
      <div
        class="cart-badge hidden"
      />
    </a>
  </nav>
  <div
    class="cart"
  >
    <h2>
      Your Cart (
      0
      )
    </h2>
    <div
      class="cart-items-container"
    />
    <div
      class="total"
    >
      <img
        alt=""
        class="total-cart-logo"
        src="shopping-cart-white.png"
      />
      $
      0.00
    </div>
  </div>
</div>
`;

exports[`Cart component renders Cart with item 1`] = `
<div>
  <nav
    class="nav-bar"
  >
    <a
      class="logo"
      href="/shopping-cart"
    >
      auditty
    </a>
    <div>
      <a
        class="nav-link"
        href="/shopping-cart"
      >
        HOME
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/music-shop"
      >
        MUSIC
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/merch-shop"
      >
        MERCH
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/sale-shop"
        style="color: red;"
      >
        SALE
      </a>
    </div>
    <a
      class="cart-logo"
      href="/shopping-cart/cart"
    >
      <img
        alt="cart"
        src="shopping-cart.png"
      />
      <div
        class="cart-badge visible"
        data-testid="cart-badge"
      >
        1
      </div>
    </a>
  </nav>
  <div
    class="cart"
  >
    <h2>
      Your Cart (
      1
      )
    </h2>
    <div
      class="cart-items-container"
    >
      <div
        class="cart-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt=""
            class="cart-item-img"
            src="mbdtf.jpg"
          />
        </a>
        <div
          class="quantity-control"
        >
          <button
            type="button"
          >
            -
          </button>
          <div
            class="quantity-label"
            data-testid="quantity-label"
          >
            1
          </div>
          <button
            type="button"
          >
            +
          </button>
          <img
            alt="remove"
            class="remove"
            src="trash.png"
          />
        </div>
        <div
          class="cart-title"
        >
          My Beautiful Dark Twisted Fantasy by Kanye West
        </div>
        <div
          class="cart-price"
        >
          $
          11.99
        </div>
      </div>
    </div>
    <div
      class="total"
    >
      <img
        alt=""
        class="total-cart-logo"
        src="shopping-cart-white.png"
      />
      $
      11.99
    </div>
  </div>
</div>
`;
